src/sentry/search/snuba/backend.py

```diff
@@ -219,43 +231,73 @@ def _query(self, project, retention_window_start, group_queryset, tags, environm
-
-        group_data = do_search(
+        candidate_hashes = None
+        if MAX_PRE_SNUBA_CANDIDATES:
+            # If the query didn't include anything to significantly filter
+            # down the number of group hashes at this point ('first_release', 'query',
+            # 'status', 'bookmarked_by', 'assigned_to', 'unassigned',
+            # 'subscribed_by', 'active_at_from', or 'active_at_to') then this
+            # queryset might return a *huge* number of groups. In this case, we
+            # probably *don't* want to pass candidates down to Snuba, and rather we
+            # want Snuba to do all the filtering/sorting it can and *then* apply
+            # this queryset to the results from Snuba.
+            #
+            # However, if this did filter down the number of groups significantly,
+            # then passing in candidates is, of course, valuable.
+
+            # {hash: group_id, ...}
+            candidate_hashes = dict(
+                GroupHash.objects.filter(
+                    group__in=group_queryset
+                ).values_list(
+                    'hash', 'group_id'
+                )[:MAX_PRE_SNUBA_CANDIDATES + 1]
+            )
+
+            if not candidate_hashes:
+                # no matches could possibly be found from this point on
+                return Paginator(Group.objects.none()).get_result()
+            elif len(candidate_hashes) > MAX_PRE_SNUBA_CANDIDATES:
+                # too many candidate hashes found, so we won't pass any down to Snuba
+                # and we'll do post-filtering instead
+                candidate_hashes = None
+
+        sort, extra_aggregations, score_fn = sort_strategies[sort_by]
+
+        # {group_id: group_score, ...}
+        snuba_groups = snuba_search(
             project_id=project.id,
             environment_id=environment and environment.id,
             tags=tags,
             start=start,
             end=end,
             sort=sort,
             extra_aggregations=extra_aggregations,
-            candidates=candidate_group_ids,
+            score_fn=score_fn,
+            candidate_hashes=candidate_hashes,
             **parameters
         )
 
-        group_to_score = {}
-        for group_id, data in group_data.items():
-            group_to_score[group_id] = calculate_cursor_for_group(data)
+        if candidate_hashes:
+            # pre-filtered candidates were passed down to Snuba,
+            # so we're finished with filtering
+            result_groups = snuba_groups.items()
+        else:
+            # pre-filtered candidates were *not* passed down to Snuba,
+            # so we need to do post-filtering to verify Sentry DB predicates
+            result_groups = []
+            for chunk in chunked(snuba_groups.items(), MAX_POST_SNUBA_CHUNK):
+                filtered_group_ids = group_queryset.filter(
+                    id__in=(gid for gid, _ in chunk)
```

Comment:
I didn't double check this to make sure it works but I'm a little surprised `__in` takes a generator like this.

Reply:
It does work, but you're right that it reads weird and just leaves a possibility of some future refactor dealing with an already consumed iterator, probably best that I just `list` it.